### PR TITLE
add stripped down schema evolution

### PR DESCRIPTION
**Pull request by hegner**
BEGINRELEASENOTES
- Adding infrastructure for schema evolution
- Added explicit version tracking to the metadata
- Data model comparison tool w/ simple heuristics to identify potential omissions / mistakes (e.g. checking for the limits of the ROOT backend)
- Changed handling of backwards compatibility for the collection info metadata
ENDRELEASENOTES

For the moment only relying on the ROOT backend. The tooling needed for SIO comes in a separate PR

### CODE REVIEW

File: python/podio_schema_evolution.py
Comment (tmadlener) [inline python/podio_schema_evolution.py:245]: Is storing these in member variables necessary? They are only used in one place, where it would be rather easy (and more legible) to just use `self.datamodel_new.components` or similar.

**Review by tmadlener — COMMENTED** [review]
I have one conceptual question: Wouldn't it be easier to store the schema version as an integer on disk? From what I can tell that would simplify quite a few things, since otherwise all comparisons would need some string manipulation first.

I have also left some smaller comments inline.

File: CMakeLists.txt
Comment (hegner) [reply CMakeLists.txt:10]: I needed it to make the tests work. As I added a version dependent read

File: include/podio/CollectionBase.h
Comment (hegner) [reply include/podio/CollectionBase.h:65]: Yes, we can easily switch to an int. I put that flexibility in because of now being able to have derived data models. And that kills the linearity in version numbers if both PODIO and other model change. 
Just a policy decision.

**Comment by tmadlener** [comment]
What is currently still missing from this? Respectively, how far do we have to take this to put more work into a follow up PR? 

I would like to merge #343 rather soonish. It doesn't lead to any merge conflicts, but it would require a few trivial fixes to the python import statements here.

File: python/podio_schema_evolution.py
Comment (tmadlener) [inline python/podio_schema_evolution.py:3]:
```suggestion
from podio.podio_config_reader import PodioConfigReader
```
fix, necessary after #343

File: src/selection.xml
Comment (tmadlener) [inline src/selection.xml:16]:
```suggestion
    <class name="std::vector<std::tuple<int, std::string, bool, unsigned>>"/>
```

**Review by tmadlener — COMMENTED** [review]
I have rebased this onto the latest version of master and fixed the few merge requests and also some of the pylint issues. I still have to go through some of them to see how to address them best.

From my point of view, the most important change from this PR is the addition of the `schemaversion` to the yaml definition and the corresponding writing of it to file.  We would need to add this writing of the schema information for the SIO backend still to this PR, but if we can get that merged, we would potentially unblock a few of the EDM4hep pull requests, because they can then for now just increment the schema version, even if we do not yet have full schema evolution capabilities.

One thing where I am not sure yet what the best way to do this is the actual generation of schema evolution code (not yet part of this PR). Here it is part of the general code generation, which has the potential to become very overloaded. Furthermore at this point it takes only one old definition file, but at some point it should take (schemaversion - 1) files to be able to get schema evolution from all old files. Also currently we take an evolution file as well, so in principle we would not need the old definition file at all, since the evolution file already should have all the information (from several versions even).

To address the latter there are two options at the moment as I see it
- Either we keep everything as it is in this PR and then iterate in "public" at the cost of a being quite unstable for our users until we have figured things out.
- Or we only keep the parts where we are more certain that they are as they should be and move the other things into a later PR where we can play a bit more with this.

There are also a few things that we need to address regardless of how we decide here (see comments below). @hegner shall I do that or do you want to have a look at them?

**Comment by hegner** [comment]
@tmadlener - did you ever look into the SIO crashes after your merge?

File: include/podio/CollectionBuffers.h
Comment (hegner) [reply include/podio/CollectionBuffers.h:47]: It depends on how we want to split up the transformation in the long run.
A possibility is to give it an additional state whether a transformation is still needed and do it in place

File: python/podio/generator_utils.py
Comment (hegner) [reply python/podio/generator_utils.py:73]: That is unfortunately not the case. I kept it separate here as chained/stacked yaml files may have different versions and thus I wanted to avoid "globals".
The default "None" we should have neither since in the future we will require a version to be there. And we want to spot the cases where we forgot,

File: python/podio_class_generator.py
Comment (hegner) [reply python/podio_class_generator.py:476]: Yes, this was a bit the blocking piece last time. Seems I forgot putting it back. :-(

File: python/podio/generator_utils.py
Comment (tmadlener) [reply python/podio/generator_utils.py:73]: I think at least for the generation part having a "global" is fine, because these are always just "global for the current EDM" that is being generated. I.e. extension models use their schema version here and there is no cross contamination from the upstream EDM.

Didn't think about catching non-defined schema versions more gracefully with the `None` default value. I think we can keep it. Maybe we should then also add a warning here to make it easier to spot from the output?

File: include/podio/CollectionBuffers.h
Comment (tmadlener) [reply include/podio/CollectionBuffers.h:47]: But doing it in place would mean to manipulate the `data` buffer, right? 

In principle in the way I imagine the actual schema evolution, you send in a `data` buffer into the black box and you get one back. Inside this black box we are free to do whatever we want, even simply nothing. However, i would like to avoid the situation where we have data in two different versions hanging around, as that is an easy way towards somewhat undefined state(s).

Comment (hegner) [reply include/podio/CollectionBuffers.h:47]: Yes, the black box approach is what I wanted. At the moment (till re-factoring) we however do not have good entities to book keep the data buffers before and after though. 
To me an item that goes into the todos

**Comment by tmadlener** [comment]
> @tmadlener - did you ever look into the SIO crashes after your merge?

I think I did, but I am not entirely sure now. I will check again.

File: python/podio_class_generator.py
Comment (tmadlener) [reply python/podio_class_generator.py:476]: Since the main concern for now is actually writing a schema version to the output files, should we defer this to later as well?

**Comment by tmadlener** [comment]
The key4hep based CI workflows are currently failing because the dynamic loading of EDM libraries picks up an old version of edm4hep. Building a consistent stack locally makes the tests run through. I will have another look into #359 again to see if we can also test this in CI before merging.

File: src/ROOTReader.cc
Comment (tmadlener) [reply src/ROOTReader.cc:196]: Especially since we have #378 now. I don't see any reason to potentially diverge on the version reported by the tag and the `CMakeLists.txt`

File: python/podio/generator_utils.py
Comment (tmadlener) [inline python/podio/generator_utils.py:77]: The roundtrip tests trip over this line here. The dumped yaml files will have `schema_version` as key in the yaml file, while the original one has `schemaversion`.

Either changing this to `schemaversion` or the yaml key to `schema_version` fixes the roundtrip tests. I would have a slight preference for renaming the yaml key to `schema_version` (mainly because it would mean less code changes).

File: tests/scripts/dumpModelRoundTrip.sh
Comment (tmadlener) [inline tests/scripts/dumpModelRoundTrip.sh:25]: This needs to stay in, because if `clang-format` is detected the original version will be clang-formatted. If no `clang-format` is found, generation will issue a warning about it, but continue.

Comment (hegner) [reply tests/scripts/dumpModelRoundTrip.sh:25]: OK. Will re-add it then.

File: python/podio/generator_utils.py
Comment (hegner) [reply python/podio/generator_utils.py:77]: I actually already went for schema_version. good that we agreed there :)

**Comment by hegner** [comment]
@tmadlener - I still have to fix the pylint messages. 
And strangely enough flake8 behaved differently on my system...

**Comment by tmadlener** [comment]
Yeah flake8 also seems to be slightly less strict locally in my case. I suspect a version mismatch is the cause for this. Here we would need some more tooling to make it easier to have the test environment available locally. But that is a general Key4hep thing, IMHO.

**Comment by jmcarcell** [comment]
I have seen that before and checked that the versions were the same as in my local pc (and the dotfiles too of course) and it was still different from the github CI :shrug: Maybe it depends on the python version but sometimes even the complains about imports are different

File: python/podio_class_generator.py
Comment (hegner) [reply python/podio_class_generator.py]: removed

**Comment by hegner** [comment]
@thomas - scheint als haengt es jetzt am aktuellen build von SIO in key4hep

File: python/podio/generator_utils.py
Comment (hegner) [reply python/podio/generator_utils.py:74]: I made it explicit now everywhere.

**Comment by tmadlener** [comment]
> @thomas - scheint als haengt es jetzt am aktuellen build von SIO in key4hep

Yes. Unfortunately not something that we can fix really easily in this case, since we would need a build with these changes in for those to disappear. It is a bit of a catch 22. However, given that we build edm4hep and run the tests there, and also that I can run these tests in a consistent environment locally, I think those should be OK to ignore.

Overall, I would like to merge #390 before this in order to make a tag before we merge this, as we haven't done one in quite some time.

**Comment by tmadlener** [comment]
Needed to resolve the conflict introduced by tagging the rest of podio

**Review by tmadlener — APPROVED** [review]

**Pull request merged**
